### PR TITLE
Master phoenix fge 2

### PR DESCRIPTION
**Pull request by FrancoisGe**

### CODE REVIEW

File: addons/html_editor/static/src/fields/html_field.js [hunk 1]
```diff
@@ -32,14 +46,66 @@ export class HtmlField extends Component {
                 this.lastValue !== record.data[this.props.name].toString()
             ) {
                 this.state.key++;
+                this.state.containsComplexHTML = this.computeContainsComplexHTML();
             }
         });
+
+        useEffect(
```
Comment (ged-odoo) [inline addons/html_editor/static/src/fields/html_field.js]: i prefer just onMounted here

Comment (ged-odoo) [reply addons/html_editor/static/src/fields/html_field.js]: no need to remove the load event actually. you can just add the listener with `once`,

File: addons/html_editor/static/src/fields/html_field.js [hunk 2]
```diff
@@ -32,14 +46,66 @@ export class HtmlField extends Component {
...
+            () => {
+                if (this.iframeRef.el) {
+                    const onLoadIframe = () => this.onLoadIframe();
+                    this.iframeRef.el.addEventListener("load", onLoadIframe);
+                    this.iframeRef.el.after(this.iframeRef.el);
```
Comment (ged-odoo) [inline addons/html_editor/static/src/fields/html_field.js]: i dont understand this

File: addons/html_editor/static/src/fields/html_field.js [hunk 3]
```diff
@@ -32,14 +46,66 @@ export class HtmlField extends Component {
...
             ) {
                 this.state.key++;
+                this.state.containsComplexHTML = this.computeContainsComplexHTML();
             }
         });
+
+        useEffect(
+            () => {
+                if (this.iframeRef.el) {
+                    const onLoadIframe = () => this.onLoadIframe();
```
Comment (ged-odoo) [inline addons/html_editor/static/src/fields/html_field.js]: maybe you can just use the onLoad from the wysiwyg component and not have to do all this manually

File: addons/html_editor/static/src/fields/html_field.js (hunk 2 above)
Comment (FrancoisGe) [reply addons/html_editor/static/src/fields/html_field.js]: I forget to had the comment from the old html_field

File: addons/html_editor/static/src/fields/html_field.js (hunk 3 above)
Comment (FrancoisGe) [reply addons/html_editor/static/src/fields/html_field.js]: It s not the wysiwyg use case. It s for the readonly mode

File: addons/html_editor/static/src/fields/html_field.js (hunk 1 above)
Comment (FrancoisGe) [reply addons/html_editor/static/src/fields/html_field.js]: > i prefer just onMounted here

Why ? It s base the record.data ?

Comment (FrancoisGe) [reply addons/html_editor/static/src/fields/html_field.js]: > no need to remove the load event actually. you can just add the listener with `once`,



> no need to remove the load event actually. you can just add the listener with `once`,

I m not sure to understand ?

File: addons/html_editor/static/src/fields/html_viewer.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    };
+
+    setup() {
+        this.iframeRef = useRef("iframe");
+
+        this.state = useState({
+            iframeVisible: false,
+        });
+
+        onMounted(() => {
```
Comment (ged-odoo) [inline addons/html_editor/static/src/fields/html_viewer.js]: suggestion:
```js
if (this.showIframe) {
          onMounted(() => {
              const onLoadIframe = () => this.onLoadIframe();
              this.iframeRef.el.addEventListener("load", onLoadIframe, { once: true });
              // Force the iframe to call the `load` event. Without this line, the
              // event 'load' might never trigger.
              this.iframeRef.el.after(this.iframeRef.el);
        });
}
```

File: addons/html_editor/static/src/fields/html_viewer.js [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        });
+
+        this.readonlyElementRef = useRef("readonlyContent");
+        useEffect(() => {
+            if (this.readonlyElementRef.el) {
+                retargetLinks(this.readonlyElementRef.el);
+            }
+        });
+
+        onWillStart(async () => {
```
Comment (ged-odoo) [inline addons/html_editor/static/src/fields/html_viewer.js]: suggestion:
```js
    if (this.props.cssAssetId) {
        onWillStart(async () => {
                this.cssAsset = await getBundle(this.props.cssAssetId);
        });
     }

File: addons/html_editor/static/src/fields/html_viewer.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                const onLoadIframe = () => this.onLoadIframe();
+                this.iframeRef.el.addEventListener("load", onLoadIframe, { once: true });
+                // Force the iframe to call the `load` event. Without this line, the
+                // event 'load' might never trigger.
+                this.iframeRef.el.after(this.iframeRef.el);
+            }
+        });
+
+        this.readonlyElementRef = useRef("readonlyContent");
+        useEffect(() => {
```
Comment (ged-odoo) [inline addons/html_editor/static/src/fields/html_viewer.js]: suggestion:
```js
     if (!this.showIframe) {
        useEffect(() => retargetLinks(this.readonlyElementRef.el));
     }
  
```

Comment (ged-odoo) [reply addons/html_editor/static/src/fields/html_viewer.js]: or rather a else from previous if

File: addons/html_editor/static/src/fields/html_field.js [hunk 7]
```diff
@@ -6,27 +6,55 @@ import { registry } from "@web/core/registry";
...
+import { HtmlViewer } from "./html_viewer";
+
+/**
+ * Check whether the current value contains nodes that would break
+ * on insertion inside an existing body.
+ *
+ * @returns {boolean} true if 'this.props.value' contains a node
+ * that can only exist once per document.
+ */
+function computeContainsComplexHTML(value) {
```
Comment (ged-odoo) [inline addons/html_editor/static/src/fields/html_field.js:18]: would it make sense to move this into htmlviewer?

Comment (FrancoisGe) [reply addons/html_editor/static/src/fields/html_field.js:18]: Actually, we need it to know if we are in readonly or not. But we can may remove it and alway pass the option sandboxPreview

File: addons/html_editor/static/src/fields/html_field.js [hunk 8]
```diff
@@ -6,27 +6,55 @@ import { registry } from "@web/core/registry";
...
+    const domParser = new DOMParser();
+    if (!value) {
```
Comment (ged-odoo) [inline addons/html_editor/static/src/fields/html_field.js:20]: you can move this above the domparser instantiation :)

File: addons/html_editor/static/src/fields/html_viewer.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    setup() {
+        this.iframeRef = useRef("iframe");
+
+        this.state = useState({
+            iframeVisible: false,
+        });
+
+        if (this.showIframe) {
+            onMounted(() => {
+                if (this.iframeRef.el) {
```
Comment (ged-odoo) [inline addons/html_editor/static/src/fields/html_viewer.js]: no need of the if here

File: addons/html_editor/static/src/fields/html_field.xml [hunk 10]
```diff
@@ -1,12 +1,15 @@
 <templates xml:space="preserve">
     <t t-name="html_editor.HtmlField">
-        <t t-if="props.readonly">
-            <div  t-ref="readonlyElement" class="o_readonly" t-out="props.record.data[this.props.name]" />
+        <t t-if="this.displayReadonly">
+            <HtmlViewer
+                value="props.record.data[props.name]"
+                cssAssetId="props.cssReadonlyAssetId"
```
Comment (ged-odoo) [inline addons/html_editor/static/src/fields/html_field.xml:6]: probably need a `t-key="'' + props.cssReadonlyAssetId + sandboxedPreview"`. If these props change, the htmlviewer will behave badly (not doing what we might expect => reloading assets and stuff)

**Pull request merged**